### PR TITLE
listener: first implementation of an Api Listener

### PR DESCRIPTION
**Pull request by junr03**
Description: this PR introduces the initial implementation of an Api Listener based on the proto configuration merged in https://github.com/envoyproxy/envoy/pull/8170. Notably, this PR introduces the ability to add only _one_ api listener via _bootstrap config only_. This decision was made in order to iterate into more complex setups (multiple listeners, LDS supplied listeners) in subsequent PRs. Moreover, the API listener is created in the context of Envoy's main thread not worker threads.

A first use of this Api Listener can be seen in https://github.com/lyft/envoy-mobile/pull/616.
Risk Level: low, only used in Envoy Mobile. The risk here is about building something generally useful and flexible. Note however that a couple of things were rejiggered in the HCM.
Testing: unit and integration tests. Additional testing in https://github.com/lyft/envoy-mobile.
Docs Changes: Added inline comments and TODOs. Proto documentation is up-to-date.
Release Notes: similar to doc changes.

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -305,6 +306,20 @@ ListenerManagerStats ListenerManagerImpl::generateStats(Stats::Scope& scope) {
 
 bool ListenerManagerImpl::addOrUpdateListener(const envoy::api::v2::Listener& config,
                                               const std::string& version_info, bool added_via_api) {
+
+  // TODO(junr03): currently only one ApiListener can be installed via bootstrap to avoid having to
+  // build a collection of listeners, and to have to be able to warm and drain the listeners. In the
+  // future allow multiple ApiListeners, and allow them to be created via LDS as well as bootstrap.
+  if (!api_listener_ && !added_via_api && config.has_api_listener()) {
```
Comment (junr03) [inline source/server/listener_manager_impl.cc]: need to figure out why the listener manager unit tests generate proto structs that _have_ an api listener. Based on my dive I can't see how to control default creation with the protobuf utils.

File: source/common/http/conn_manager_impl.cc [hunk 2]
```diff
@@ -285,6 +285,11 @@ void ConnectionManagerImpl::handleCodecException(const char* error) {
   read_callbacks_->connection().close(Network::ConnectionCloseType::FlushWriteAndDelay);
 }
 
+void ConnectionManagerImpl::forceCodecCreation() {
```
Comment (goaway) [inline source/common/http/conn_manager_impl.cc]: Maybe just call this `createCodec`, have it take data as an argument, and use it internally as well (e.g. in `onData` below).

File: source/extensions/filters/network/http_connection_manager/config.h [hunk 3]
```diff
@@ -199,6 +200,17 @@ class HttpConnectionManagerConfig : Logger::Loggable<Logger::Id::config>,
   static const uint64_t RequestTimeoutMs = 0;
 };
 
+/**
+ * Factory to create an HttpConnectionManager outside of a Network Filter Chain.
+ */
+class HttpConnectionManagerFactory {
+public:
+  static std::function<Http::ServerConnectionCallbacksPtr()>
+  createHttpConnectionManagerFactoryFromProto(const ProtobufWkt::Any& proto_config,
```
Comment (goaway) [inline source/extensions/filters/network/http_connection_manager/config.h]: maybe `createStandaloneHttpConnectionManagerFactory`?

**Comment by junr03** [comment]
@mattklein123 @goaway I'd love some eyes on this if you have some time today.

**Comment by mattklein123** [comment]
Can you merge master and then I will review? Thank you.

/wait

**Comment by junr03** [comment]
merged master. Thanks for reviewing @mattklein123

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 4]
```diff
@@ -498,6 +498,74 @@ const Network::Address::Instance& HttpConnectionManagerConfig::localAddress() {
   return *context_.localInfo().address();
 }
 
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(hcm_date_provider);
+SINGLETON_MANAGER_REGISTRATION(hcm_route_config_provider_manager);
+SINGLETON_MANAGER_REGISTRATION(hcm_scoped_routes_config_provider_manager);
+
+// TODO(junr03): some of this code can be shared with the other factory code. Clean up
+// if this factory approach is well received by reviewers.
```
Comment (mattklein123) [inline source/extensions/filters/network/http_connection_manager/config.cc]: Yeah this seems fine, so would like to see the cleaned up shared version

File: source/extensions/filters/network/http_connection_manager/config.h (hunk 3 above)
Comment (mattklein123) [reply source/extensions/filters/network/http_connection_manager/config.h]: I'm fine either way

File: source/server/api_listener_impl.h [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    // Synthetic class that acts as a stub for the connection backing the
+    // Network::ReadFilterCallbacks.
+    class SyntheticConnection : public Network::Connection {
+    public:
+      SyntheticConnection(SyntheticReadCallbacks& parent)
+          : parent_(parent), stream_info_(parent_.parent_.timeSource()),
+            options_(std::make_shared<std::vector<Network::Socket::OptionConstSharedPtr>>()) {}
+
+      // Network::FilterManager
+      void addWriteFilter(Network::WriteFilterSharedPtr) override {}
```
Comment (mattklein123) [inline source/server/api_listener_impl.h]: I would probably NOT_IMPLEMENTED all of these that you never expect to get called for clarity. Obviously this is a bit of an interface smell but I'm not sure there is much we can easily do about that.

File: source/server/api_listener_impl.h [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    HttpApiListenerImpl& parent_;
+    SyntheticConnection connection_;
+  };
+
+  const envoy::config::listener::v3alpha::Listener& config_;
+  ListenerManagerImpl& parent_;
+  const std::string name_;
+  Network::Address::InstanceConstSharedPtr address_;
+  ProtobufMessage::ValidationVisitor& validation_visitor_;
+  Stats::ScopePtr global_scope_;
```
Comment (mattklein123) [inline source/server/api_listener_impl.h]: Why is this needed? Can't the global scope be used directly? Or is this to match the existing listeners?

File: source/common/http/conn_manager_impl.cc (hunk 2 above)
Comment (junr03) [reply source/common/http/conn_manager_impl.cc]: Given right now there is only one I was going to leave this to a subsequent PR that introduced the ability to have multiple API listeners. Otherwise, I couldn't think of what other stats to expose at the moment.

File: source/server/api_listener_impl.h [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  // that rather than this semi-real one.
+  const Network::Address::InstanceConstSharedPtr& address() const { return address_; }
+
+  // ApiListener
+  absl::string_view name() const override { return name_; }
+  ApiListenerHandle* handle() override;
+
+  // TODO(junr03): the majority of this surface could be moved out of the listener via some sort of
+  // base class context.
+  // Server::Configuration::FactoryContext
```
[the remark this answers is not in the document]
Comment (junr03) [reply source/server/api_listener_impl.h]: will take a look, thanks for linking.

File: source/server/listener_impl.h [hunk 8]
```diff
@@ -4,6 +4,7 @@
 
 #include "envoy/config/core/v3alpha/base.pb.h"
 #include "envoy/config/listener/v3alpha/listener.pb.h"
+#include "envoy/network/connection.h"
```
[the remark this answers is not in the document]
Comment (junr03) [reply source/server/listener_impl.h]: ah left behind from when I initially wrote the api listener class in here. Will delete

File: source/server/listener_manager_impl.cc [hunk 9]
```diff
@@ -309,6 +310,23 @@ ListenerManagerStats ListenerManagerImpl::generateStats(Stats::Scope& scope) {
 bool ListenerManagerImpl::addOrUpdateListener(
     const envoy::config::listener::v3alpha::Listener& config, const std::string& version_info,
     bool added_via_api) {
+
+  // TODO(junr03): currently only one ApiListener can be installed via bootstrap to avoid having to
+  // build a collection of listeners, and to have to be able to warm and drain the listeners. In the
+  // future allow multiple ApiListeners, and allow them to be created via LDS as well as bootstrap.
```
[the remark this answers is not in the document]
Comment (junr03) [reply source/server/listener_manager_impl.cc]: Exactly. That was the first approach that I tackled this with. I spent a few hours toying with different implementations. However, there was a good amount of refactoring required to get us there. Mostly around the API surface that the listener manager currently uses to interact with the listener, and how most of it was not applicable to API listeners. This meant that I needed to start messing more than I wanted with the listener manager code.

So the conclusion that I arrived at was this simple approach to get something in place. I agree that we can then build up to the Listener manager owning an interface rather than the implementations.

File: source/server/api_listener_impl.h (hunk 5 above)
Comment (junr03) [reply source/server/api_listener_impl.h]: Yeah I thought about this for a bit, but without refactoring to separate the filter management aspect from the listener aspect per se, I didn't think we could do anything to clean up the interface.

I'll go ahead and replace these with NOT_IMPLEMENTED.

File: source/server/api_listener_impl.h (hunk 6 above)
Comment (junr03) [reply source/server/api_listener_impl.h]: yep, I was just matching the existing listener here.

File: include/envoy/server/api_listener.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Handle provided to users to interact with Envoy via a particular API.
+ * Given the flexibility of the api_listener config, this interface is also maximally flexible.
+ * Consumers of this interface have to cast this handle to narrower types.
+ */
+class ApiListenerHandle {
```
[the remark this answers is not in the document]
Comment (junr03) [reply include/envoy/server/api_listener.h]: I like the return patter in address, nice to avoid the dynamic cast. I went ahead and did it. 

I made the ConnectionManagerImpl itself implement Http::ApiListener, but I could also see how it would make sense for the derived HttpApiListener class to implement itself and dispatch the newStream call. There are a couple ways to organize it, let me know what you think.

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 4 above)
Comment (junr03) [reply source/extensions/filters/network/http_connection_manager/config.cc]: updated

**Comment by junr03** [comment]
this is ready for another pass. Thanks for all the comments @mattklein123 and @goaway

File: api/envoy/config/listener/v2/api_listener.proto [hunk 11]
```diff
@@ -8,13 +8,12 @@ option java_package = "io.envoyproxy.envoy.config.listener.v2";
...
 option java_multiple_files = true;
 
-// [#not-implemented-hide:]
 // Describes a type of API listener, which is used in non-proxy clients. The type of API
 // exposed to the non-proxy application depends on the type of API listener.
 message ApiListener {
   // The type in this field determines the type of API listener. At present, the following
   // types are supported:
-  //   envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager (HTTP)
+  // envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager (HTTP)
```
Comment (mattklein123) [inline api/envoy/config/listener/v2/api_listener.proto]: @htuch do you have any thoughts on how this type of documentation could be automatically kept up to date in the future? I'm wondering if we could somehow if a protodoc insertion point that inserts all known variants of a message or something like that. Not sure what the best thing to do here is. This will need to be kept up to date for v3 very soon for example.

File: include/envoy/http/api_listener.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  /**
+   * Invoked when a new request stream is initiated by the remote.
+   * @param response_encoder supplies the encoder to use for creating the response. The request and
+   *                         response are backed by the same Stream object.
+   * @param is_internally_created indicates if this stream was originated by a
+   *   client, or was created by Envoy, by example as part of an internal redirect.
+   * @return StreamDecoder& supplies the decoder callbacks to fire into for stream decoding events.
+   */
+  virtual StreamDecoder& newStream(StreamEncoder& response_encoder,
+                                   bool is_internally_created = false) PURE;
```
Comment (mattklein123) [inline include/envoy/http/api_listener.h]: In a perfect world we would avoid duplicating this interface method in multiple places. You might be able to avoid this with a slight rejiggering of inheritance. I would at least leave a TODO around this.

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 13]
```diff
@@ -78,11 +75,10 @@ SINGLETON_MANAGER_REGISTRATION(date_provider);
...
 SINGLETON_MANAGER_REGISTRATION(scoped_routes_config_provider_manager);
 
-Network::FilterFactoryCb
-HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped(
-    const envoy::extensions::filters::network::http_connection_manager::v3alpha::
-        HttpConnectionManager& proto_config,
-    Server::Configuration::FactoryContext& context) {
+std::tuple<std::shared_ptr<Http::TlsCachingDateProviderImpl>,
+           std::shared_ptr<Router::RouteConfigProviderManager>,
+           std::shared_ptr<Router::ScopedRoutesConfigProviderManager>>
```
Comment (mattklein123) [inline source/extensions/filters/network/http_connection_manager/config.cc]: nit: I would just return a struct with 3 members. It is easier to read and will be easy to capture in the lambda, etc.

File: api/envoy/config/listener/v2/api_listener.proto (hunk 11 above)
Comment (htuch) [reply api/envoy/config/listener/v2/api_listener.proto]: The RST anchors are automagically transformed via `protoxform`. I think we should add an annotation to support this, so that `protoxform` can (1) do consistency checking to ensure there aren't typos and (2) automatic upgrade.

File: source/server/api_listener_impl.cc [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+Upstream::ClusterManager& ApiListenerImpl::clusterManager() {
+  return parent_.server_.clusterManager();
+}
+Event::Dispatcher& ApiListenerImpl::dispatcher() { return parent_.server_.dispatcher(); }
+Network::DrainDecision& ApiListenerImpl::drainDecision() { return *this; }
+Grpc::Context& ApiListenerImpl::grpcContext() { return parent_.server_.grpcContext(); }
+bool ApiListenerImpl::healthCheckFailed() { return parent_.server_.healthCheckFailed(); }
+Tracing::HttpTracer& ApiListenerImpl::httpTracer() { return httpContext().tracer(); }
+Http::Context& ApiListenerImpl::httpContext() { return parent_.server_.httpContext(); }
+Init::Manager& ApiListenerImpl::initManager() { return parent_.server_.initManager(); }
```
Comment (lambdai) [inline source/server/api_listener_impl.cc]: drive by from https://github.com/envoyproxy/envoy/pull/9205#discussion_r363866697
I am thinking about adding a FilterChainFactoryContext without a ListenerImpl and InitManager is the first concern.
While looking at your impl here, I feel this is probably a glitch.
You are reusing the server's init manager. It is guaranteed to be initialized after worker thread started(search for ListenerManagerImpl::workers_started_)
So if api listener is added through LDS *after* server is initialized the init manager the init manager here doesn't HttpConnManager.

Comment (junr03) [reply source/server/api_listener_impl.cc]: Right now the ApiListener can only be created during initial bootstrap so that is why we are only using the server's init manager. I plan on adding dynamic ApiListeners in a subsequent PR where, as you say, the init manager will have to change to a local one like in the ListenerImpl.

File: source/server/api_listener_impl.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (junr03) [reply source/server/api_listener_impl.h]: I synced up with @lambdai yesterday afternoon. I updated the todo with our discussion to update in a subsequent PR.

**Comment by junr03** [comment]
@mattklein123 @goaway this is ready for another pass.

Sorry for the lag, I was dealing with a couple other items the last couple of days.

I think this is functionally complete as a first step. The PR has TODOs around the following areas:
- Clean up: ApiListener interface duplication with the ServerConnectionCallbacks,  FactoryContext base class, potentially separate L4 responsibilities of the HCM to prevent synthetic conn/read callbacks, have API listeners not require an address, allow the HCM to have a synthetic codec.
- Additional improvements to API listeners: allow more that one listener, allow listeners to be dynamically added, allow building other concrete implementations.
- TODOs for code that was there: separate build targets for the listener manager and the listener libs. This might be hard because of a circular dependency. But I felt that at least calling out in a TODO was valuable.

This TODO work can be done iteratively and in parallel of this getting merged in support of https://github.com/lyft/envoy-mobile/pull/616.

**Comment by junr03** [comment]
looking at the failed CI jobs

File: include/envoy/server/listener_manager.h [hunk 15]
```diff
@@ -212,6 +213,14 @@ class ListenerManager {
...
   using FailureStates = std::vector<std::unique_ptr<envoy::admin::v3alpha::UpdateFailureState>>;
   virtual void endListenerUpdate(FailureStates&& failure_states) PURE;
+
+  // TODO(junr03): once ApiListeners support warming and draining, this function should return a
+  // weak_ptr to its caller. This would allow the caller to verify if the
+  // ApiListener is available to receive API calls on it.
+  /**
+   * @return ApiListener* the server's API Listener if it exists, nullptr if it does not.
+   */
+  virtual ApiListener* apiListener() PURE;
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: nit: I would use a type aliased optional reference wrapper here.

File: source/server/api_listener_impl.h [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+  // that rather than this semi-real one.
+  const Network::Address::InstanceConstSharedPtr& address() const { return address_; }
+
+  // ApiListener
+  absl::string_view name() const override { return name_; }
+
+  // TODO(junr03): move these functions into a concrete Server:Configuration::FactoryContext that
+  // takes the server's context, and the listener config in the constructor. The only challenging
+  // part is implementing the initManager() call. The function depends on the Listener's state for
+  // ListenerImpl, and will do so once ApiListeners are dynamic.
```
Comment (mattklein123) [inline source/server/api_listener_impl.h]: Can we just do this now with the utility class that @lambdai added? It's already basically a wrapper and it seems not difficult to share the wrapper in some way with this code.

File: source/server/listener_manager_impl.cc [hunk 17]
```diff
@@ -319,6 +320,25 @@ ListenerManagerStats ListenerManagerImpl::generateStats(Stats::Scope& scope) {
...
+    if (!api_listener_ && !added_via_api) {
+      // TODO(junr03): dispatch to different concrete constructors when there are other
+      // ApiListenerImpl derived classes.
+      api_listener_ = std::make_unique<HttpApiListener>(
+          config, *this, config.name(),
+          server_.messageValidationContext().staticValidationVisitor());
+      return true;
+    } else {
+      ENVOY_LOG(debug, "listener {} can not be added because currently only one ApiListener is "
+                       "allowed, and it can only be added via bootstrap configuration");
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: This should probably be a warning?

File: source/common/http/conn_manager_impl.cc [hunk 18]
```diff
@@ -286,18 +286,26 @@ void ConnectionManagerImpl::handleCodecException(const char* error) {
...
+  if (codec_->protocol() == Protocol::Http3) {
+    stats_.named_.downstream_cx_http3_total_.inc();
+    stats_.named_.downstream_cx_http3_active_.inc();
+  } else if (codec_->protocol() == Protocol::Http2) {
+    stats_.named_.downstream_cx_http2_total_.inc();
+    stats_.named_.downstream_cx_http2_active_.inc();
+  } else {
+    stats_.named_.downstream_cx_http1_total_.inc();
+    stats_.named_.downstream_cx_http1_active_.inc();
+  }
```
Comment (goaway) [inline source/common/http/conn_manager_impl.cc]: nit: since there's 3 options now, maybe `switch`.  Also could consider `ASSERT`/`PANIC` on `default:` (indicating something was missed).

File: include/envoy/server/listener_manager.h (hunk 15 above)
Comment (junr03) [reply include/envoy/server/listener_manager.h]: updated. Out of curiosity do you prefer these because the clarity that the caller does not own the object, or because optional "forces" you to think about nullopt, or something else?

Also there is no consistent naming for these type aliases. I went for the one with precedent that seemed closest to the smart pointer naming convention. I can homogenize naming for these aliases in a follow up PR.

Comment (mattklein123) [reply include/envoy/server/listener_manager.h]: Yes because it's clear that it's not a raw pointer that can be deleted.

File: source/server/api_listener_impl.h (hunk 16 above)
Comment (junr03) [reply source/server/api_listener_impl.h]: I went ahead and created an object here. I don't think there is a clear way to use the wrapper pattern and not have the listener implement the FactoryContext interface, because someone has to implement that interface. i.e, someone has to be the inner wrapped object. Let me know if you see an alternative I am missing here.

**Comment by junr03** [comment]
@mattklein123 ready for another pass. The integration test was pretty straightforward to write. I have some open questions about the FactoryContext, let me know what you think.

File: source/server/api_listener_impl.h (hunk 16 above)
Comment (junr03) [reply source/server/api_listener_impl.h]: cc @lambdai if you want to take a look

Comment (lambdai) [reply source/server/api_listener_impl.h]: To me the FilterChainContext should delegate functionality to another (shared) object to avoid duplicating the code/members in replicas of FilterChainContext.
My understanding the shared object is listener_impl/api_listener_impl now, and could be future wrapped thing if we unify some code between plain listener_impl and api_listener_impl

Comment (lambdai) [reply source/server/api_listener_impl.h]: And the unify could be an independent PR, as long as the impl of api listener is not immediately abused by elsewhere

**Comment by mattklein123** [comment]
Sorry needs master merge.

/wait

**Comment by junr03** [comment]
@mattklein123 merged!

File: include/envoy/server/api_listener.h [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+   * @return the Type of the ApiListener.
+   */
+  virtual Type type() const PURE;
+
+  /**
+   * @return Http::ApiListener IFF type() == Type::HttpApiListener, otherwise nullptr.
+   */
+  virtual Http::ApiListener* http() PURE;
```
Comment (mattklein123) [inline include/envoy/server/api_listener.h]: nit: prefer something similar to ApiListenerOptRef here also as the return value.

File: include/envoy/server/listener_manager.h [hunk 20]
```diff
@@ -212,6 +213,14 @@ class ListenerManager {
    */
   using FailureStates = std::vector<std::unique_ptr<envoy::admin::v3::UpdateFailureState>>;
   virtual void endListenerUpdate(FailureStates&& failure_states) PURE;
+
+  // TODO(junr03): once ApiListeners support warming and draining, this function should return a
+  // weak_ptr to its caller. This would allow the caller to verify if the
+  // ApiListener is available to receive API calls on it.
+  /**
+   * @return ApiListener* the server's API Listener if it exists, nullptr if it does not.
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: nit: return type wrong, or just remove

File: source/common/http/conn_manager_impl.cc [hunk 21]
```diff
@@ -286,18 +286,33 @@ void ConnectionManagerImpl::handleCodecException(const char* error) {
...
+    stats_.named_.downstream_cx_http2_total_.inc();
+    stats_.named_.downstream_cx_http2_active_.inc();
+    break;
+  case Protocol::Http11:
+  case Protocol::Http10:
+    stats_.named_.downstream_cx_http1_total_.inc();
+    stats_.named_.downstream_cx_http1_active_.inc();
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc]: Is this actually required? The compiler should check all enum values are covered.

File: source/server/api_listener_impl.cc [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+      listener_scope_(parent_.server_.stats().createScope(fmt::format("listener.api.{}.", name_))),
+      factory_context_(parent_.server_, config_, *this, *global_scope_, *listener_scope_),
+      read_callbacks_(SyntheticReadCallbacks(*this)) {}
+
+HttpApiListener::HttpApiListener(const envoy::config::listener::v3::Listener& config,
+                                 ListenerManagerImpl& parent, const std::string& name)
+    : ApiListenerImpl(config, parent, name) {
+  auto typed_config = MessageUtil::anyConvert<
+      envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager>(
+      config.api_listener().api_listener());
```
Comment (mattklein123) [inline source/server/api_listener_impl.cc]: I think you need to validate after calling anyConvert, please check this. It might be useful to add an anyConvertAndValidate variant and replace that where relevant.

File: test/integration/api_listener_integration_test.cc [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::ReturnRef;
+
+namespace Envoy {
+namespace {
+
+class ApiListenerIntegrationTest : public BaseIntegrationTest,
+                                   public testing::TestWithParam<Network::Address::IpVersion> {
```
Comment (mattklein123) [inline test/integration/api_listener_integration_test.cc]: Can this potentially be based on the HTTP integration tests? I think you could just use that with a config modifier that replaces the listener contents with an API listener?

File: test/integration/api_listener_integration_test.cc [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+TEST_P(ApiListenerIntegrationTest, Basic) {
+  ConditionalInitializer test_ran;
+  test_server_->server().dispatcher().post([this, &test_ran]() -> void {
+    ASSERT_TRUE(test_server_->server().listenerManager().apiListener().has_value());
+    ASSERT_EQ("api_listener", test_server_->server().listenerManager().apiListener()->get().name());
+    ASSERT_NE(nullptr, test_server_->server().listenerManager().apiListener()->get().http());
+    auto* http_api_listener = test_server_->server().listenerManager().apiListener()->get().http();
+
+    ON_CALL(stream_encoder_, getStream()).WillByDefault(ReturnRef(stream_encoder_.stream_));
+    auto& stream_decoder = http_api_listener->newStream(stream_encoder_);
```
Comment (mattklein123) [inline test/integration/api_listener_integration_test.cc]: Why are we using mocks here? Can't these be a real end to end tests with an usptream?

Comment (junr03) [reply test/integration/api_listener_integration_test.cc]: I am mocking the client bits to set expectations on what the client will receive. I am not sure what real class would make sense to use here otherwise. Unless I create a small Stream/StreamEncoder class?

For instance, in the Envoy Mobile integration tests I will use the actual Http::DirectStream and the Http::DirectStreamCallbacks and assert on values received. But here is made sense to me to mock the client and have expectations.

File: include/envoy/server/api_listener.h (hunk 19 above)
Comment (junr03) [reply include/envoy/server/api_listener.h]: yeah will convert. Given it is new code might as well use the pattern where appropriate.

File: test/integration/api_listener_integration_test.cc (hunk 24 above)
Comment (mattklein123) [reply test/integration/api_listener_integration_test.cc]: Ah OK, I see. Alright that seems fine, thanks.

File: source/server/api_listener_impl.cc (hunk 22 above)
Comment (junr03) [reply source/server/api_listener_impl.cc]: Good catch, I created the function. There are several sites where the function call be updated. I'll take that in a separate PR https://github.com/envoyproxy/envoy/issues/9712

**Comment by junr03** [comment]
@mattklein123 updated! Thanks for all the comments :)
